### DISCUSSION TITLE
Multiple Assignment with one declared, one undeclared variable

### DISCUSSION THREAD

**Discussion by ruipsrosario** [odin-lang/Odin#959] [Q&A] [answered]
Hey there,

I was coding some stuff with Odin and I am currently facing this "problem" (not sure if it is by design or not).

So imagine you have procedures `A` and `B` which both return `(AType, string)` and `(AnotherType, string)`, respectively.  And you have this snippet of code:

    aType, errorString := A();
    anotherType, errorString := B();

The compiler will complain since the call to procedure `B` will attempt to redeclare `errorString`. However, the intent was to reuse `errorString`, from the declaration of calling `A` earlier, and declare `anotherType`. Due to using `:=` the compiler will interpret as declaring _both_ variables. Using `=` the compiler will complain saying that `anotherType` isn't defined.

**Is there going to be support to declaring only some variables in a multiple return values procedure call?** I know this can be fixed by just declaring all variables earlier, but since Odin does a good job at reducing LOC this could further help reducing it.

Cheers,
Rui Rosário

_example code, not code from actual situation_

**Comment by alichay** [comment]
I'm under the impression that something like that would serve to complicate
the language more than it needs to be.

For my code, when I run into that, I just forward declare the variables,
then use the assignment operator. It works out fine and isn't really any
more complex.

**Comment by ruipsrosario** [comment]
@zangent I don't mind forward declaring variables (I'm already used to that), I was just curious if it was something that was planned but overlooked or just something that was ignored by design.

**Comment by gingerBill** [comment, marked as answer]
TL;DR This is by design and not a mistake.

`:=` Is actually `:  =` with the type missing. So what you are doing is double declaration that variable.
```
// These three are equivalent in semantics
x: int; x = 123;
x: int = 123;
x := 123;
```
If you do have a "tuple" value on the right hand side, you either have to predeclare all the variables or remove the type (or specify the type if they are all of the same type). Tuples are not a first class type in this language and this is by design. I've found that tuples are only ever useful as return values from procedures and a named membered record is a much better thing to have (even if it is a little more verbose).

In Go, `:=` is called a short variable declaration which will reassign variables that have previously been declared. I personally do not like this approach as it does hidden variable reassignment which is very prone to bugs (which I've had happen to me).

I hope this clears things up. You can still use `:=` but I would either recommend declaring variables beforehand or rename that new variable.
